File: 0129-sum-root-to-leaf-numbers/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void sumNumbers(TreeNode* root, int & curSum, int& totalSum)
    {
        if(root == nullptr)
        {
            return;
        }
        else if(root->left == nullptr && root->right == nullptr)
        {
            curSum = curSum*10 + root->val;
            totalSum += curSum;
            return;
        }
        
        int copy = curSum;
        
        curSum = curSum*10 + root->val;
        sumNumbers(root->left, curSum, totalSum);
        
        curSum = copy*10 + root->val;
        sumNumbers(root->right, curSum, totalSum);
    }
    int sumNumbers(TreeNode* root) 
    {
        int curSum, totalSum;
        curSum = totalSum = 0;
        sumNumbers(root, curSum, totalSum);
        return totalSum;
    }
};

File: 0129-sum-root-to-leaf-numbers/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def sumNumbers(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        root_to_leaf = cur_num = 0
        
        while root:
            
            if root.left:
                parent = root.left
                steps       = 1
                
                while parent.right and parent.right is not root:
                    parent = parent.right
                    steps  += 1
                
                if parent.right is None:
                    cur_num = cur_num*10 + root.val
                    parent.right = root
                    root         = root.left
                    
                else:
                    if parent.left is None:
                        root_to_leaf += cur_num
                    cur_num = cur_num//pow(10, steps)
                    parent.right = None
                    root = root.right
            else:
                cur_num = cur_num*10 + root.val
                if root.right is None:
                    root_to_leaf += cur_num
                
                root = root.right
        return root_to_leaf
                
        
